### PR TITLE
[WIP] CPU_FLAGS_ARM

### PR DESCRIPTION
**Pull request by mgorny**
Done so far:
- [x] .desc and USE_EXPAND for CPU_FLAGS_ARM
- [x] profile masks/forces
- [ ] packages
- [ ] cleanup of old flags from profiles

@aballier, @leio, @gentoo/arm

Help would much appreciated.

@aballier, could you please convert ffmpeg and libav? I don't want to figure out those arrays ;-).

Could also someone shed some light on why the profiles are doing what they're doing? Like why most of the flags are masked on most of arch profiles, and e.g. armv5te profile unmasks (but does not force) armv5te?

### CODE REVIEW

**Comment by mgorny** [comment]
(note that you can push to the branch on my gh)

**Comment by mgorny** [comment]
@aballier, @leio, @gentoo/arm: please review and test this version. This time I've only added masks/forces for new flags, without touching the old ones. If this is right, we can start migrating stuff.

File: profiles/desc/cpu_flags_arm.desc [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+sha1 - Use the SHA-1 cryptography instruction set
+sha2 - Use the SHA-2 cryptography instruction set
+thumb2 - Enable Thumb-2 instruction set (ARMv6T2 and ARMv7+)
+v4 - Use ARMv4 instructions
+v4t - Use ARMv4T instructions
+v5t - Use ARMv5T instructions
+v5te - Use ARMv5TE instructions
+v5tej - Use ARMv5TEJ instructions
+v6 - Use ARMv6 instructions
+v7 - Use ARMv7 instructions
```
Comment (aballier) [inline profiles/desc/cpu_flags_arm.desc]: v8 too maybe ?

File: profiles/desc/cpu_flags_arm.desc [hunk 2]
```diff
@@ -0,0 +1,22 @@
+# Copyright 1999-2017 Gentoo Foundation.
+# Distributed under the terms of the GNU General Public License v2
+
+aes - Use the AES cryptography instruction set
+crc32 - Use the CRC32 instruction set
+iwmmxt - Use the iwMMXt instruction set
+neon - Use the NEON (ASIMD) instruction set
+sha1 - Use the SHA-1 cryptography instruction set
+sha2 - Use the SHA-2 cryptography instruction set
+thumb2 - Enable Thumb-2 instruction set (ARMv6T2 and ARMv7+)
```
Comment (aballier) [inline profiles/desc/cpu_flags_arm.desc]: Not sure if you mean to be complete or just add them as-needed, there would also be thumb and thumbee:
http://infocenter.arm.com/help/index.jsp?topic=/com.arm.doc.dui0473f/CEGBEIJB.html

File: profiles/desc/cpu_flags_arm.desc [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+v4 - Use ARMv4 instructions
+v4t - Use ARMv4T instructions
+v5t - Use ARMv5T instructions
+v5te - Use ARMv5TE instructions
+v5tej - Use ARMv5TEJ instructions
```
Comment (aballier) [inline profiles/desc/cpu_flags_arm.desc]: Can't find an authoritative document, but maybe name this "Jazelle":
https://en.wikipedia.org/wiki/Jazelle

Jazelle is denoted by a 'J' appended to the CPU name, except for post-v5 cores where it is required (albeit only in trivial form) for architecture conformance.

File: profiles/desc/cpu_flags_arm.desc [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+aes - Use the AES cryptography instruction set
+crc32 - Use the CRC32 instruction set
+iwmmxt - Use the iwMMXt instruction set
+neon - Use the NEON (ASIMD) instruction set
+sha1 - Use the SHA-1 cryptography instruction set
+sha2 - Use the SHA-2 cryptography instruction set
+thumb2 - Enable Thumb-2 instruction set (ARMv6T2 and ARMv7+)
+v4 - Use ARMv4 instructions
+v4t - Use ARMv4T instructions
+v5t - Use ARMv5T instructions
```
Comment (aballier) [inline profiles/desc/cpu_flags_arm.desc]: With a thumb useflag you could drop the t variants:
https://www.altera.com/en_US/pdfs/literature/third-party/archives/ddi0100e_arm_arm.pdf

Comment (aballier) [inline profiles/desc/cpu_flags_arm.desc]: To name a precise version and variant of the ARM/Thumb architecture, the following strings are concatenated:
The string ARMv.
The version number of the ARM instruction set.
Variant letters of the included variants, except that the M variant is standard in architecture versions 4 and above, and therefore not normally listed.
If any variants described as standard in 3 above are not present, the letter x followed by the letters of the excluded variants. In addition, the letter P can be used after x to denote the exclusion of certain instructions from architecture version ARMv5TE, as described in The ARMv5TExP architecture version.

File: profiles/desc/cpu_flags_arm.desc [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+v5te - Use ARMv5TE instructions
```
Comment (aballier) [inline profiles/desc/cpu_flags_arm.desc]: The 'e' here could be dropped too and replaced by a kind of 'edsp' useflag

File: profiles/desc/cpu_flags_arm.desc (hunk 2 above)
Comment (mgorny) [inline profiles/desc/cpu_flags_arm.desc]: I'd rather add them as-needed. Wasn't `thumbee` deprecated anyway? I think I've seen such an info somewhere.

File: profiles/desc/cpu_flags_arm.desc (hunk 5 above)
Comment (mgorny) [inline profiles/desc/cpu_flags_arm.desc]: How is edsp related to vfp?

File: profiles/desc/cpu_flags_arm.desc (hunk 3 above)
Comment (mgorny) [inline profiles/desc/cpu_flags_arm.desc]: So maybe drop it altogether, I don't think anything explicitly needs -j variant.

File: profiles/desc/cpu_flags_arm.desc (hunk 1 above)
Comment (mgorny) [inline profiles/desc/cpu_flags_arm.desc]: I wasn't sure about that. Probably makes sense.

File: profiles/desc/cpu_flags_arm.desc (hunk 2 above)
Comment (aballier) [inline profiles/desc/cpu_flags_arm.desc]: yes I've seen it too

thumb allows to drop some variants below so probably worth it though

File: profiles/desc/cpu_flags_arm.desc (hunk 5 above)
Comment (aballier) [inline profiles/desc/cpu_flags_arm.desc]: Doesn't seem related:
 VFPv2 is an optional extension to the ARM instruction set
in the ARMv5TE, ARMv5TEJ and ARMv6 architectures

File: profiles/desc/cpu_flags_arm.desc (hunk 3 above)
Comment (aballier) [inline profiles/desc/cpu_flags_arm.desc]: probably

**Comment by mgorny** [comment]
@aballier, @steev: I've updated cpu_flags_arm.desc, please re-review.

@steev also mentioned that NEON in aarch64 is (sometimes?) incompatible with the one for 32-bit arm. Do we need to account for that somehow? The cpuinfo flag is different too in the kernel.

**Comment by aballier** [comment]
lgtm now :+1:

File: profiles/arch/arm/armv5te/use.force [hunk 6]
```diff
@@ -0,0 +1,5 @@
+# Force armv5 specific USE flags
+cpu_flags_arm_v4
+cpu_flags_arm_v4t
+cpu_flags_arm_v5t
+cpu_flags_arm_v5te
```
Comment (aballier) [inline profiles/arch/arm/armv5te/use.force]: the whole patch needs updating here i think

**Comment by mgorny** [comment]
Ok, now I've updated it all + cpuid2cpuflags arm-wip branch.

File: profiles/desc/cpu_flags_arm.desc [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+neon - Use the NEON (ASIMD) instruction set
+sha1 - Use the SHA-1 cryptography instruction set
+sha2 - Use the SHA-2 cryptography instruction set
+thumb - Enable Thumb instruction set (ARMv*T and ARMv6+)
+thumb2 - Enable Thumb-2 instruction set (ARMv*T2 and ARMv7+)
+v4 - Use ARMv4 instructions
+v6 - Use ARMv6 instructions
+v7 - Use ARMv7 instructions
+v8 - Use ARMv8 instructions
+vfp - Use the VFP instruction set
```
Comment (aballier) [inline profiles/desc/cpu_flags_arm.desc]: https://www.arm.com/products/processors/technologies/vector-floating-point.php

Which one is it? v1 or v2?

File: profiles/desc/cpu_flags_arm.desc [hunk 8]
```diff
@@ -0,0 +1,21 @@
...
+vfp-d16 - Indicate that the FPU has 16 64-bit VFP registers
+vfp-d32 - Indicate that the FPU has 32 64-bit VFP registers
+vfpv3 - Use the VFP version 3 instruction set
```
Comment (aballier) [inline profiles/desc/cpu_flags_arm.desc]: Based on https://www.arm.com/products/processors/technologies/vector-floating-point.php
I'd merge vfp-d16 with vfpv3

VFPv3 is an optional extension to the ARM, Thumb® and ThumbEE instruction sets in the ARMv7-A and ARMv7-R profiles. VFPv3 implementation is with either thirty-two or sixteen double word registers. The terms VFPv3-D32 and VFPv3-D16 distinguish between these two implementation options.

File: profiles/arch/arm/armv5te/use.mask [hunk 9]
```diff
@@ -1,2 +1,10 @@
 # Unmask armv5 specific USE flags
 -armv5te
+
+-cpu_flags_arm_v4
+-cpu_flags_arm_v5
+
+-cpu_flags_arm_edsp
+-cpu_flags_arm_thumb
+-cpu_flags_arm_vfp
+-cpu_flags_arm_vfp-d16
```
Comment (aballier) [inline profiles/arch/arm/armv5te/use.mask]: Since we don't support anything pre armv5, I'd assume vfp means vfpv2

vfp-d16 seems to be for 16 double registers vfpv3 on armv7, so should be masked here I think

File: profiles/arch/arm/armv6j/use.force [hunk 10]
```diff
@@ -0,0 +1,10 @@
+# Force armv6 specific USE flags
+cpu_flags_arm_v4
+cpu_flags_arm_v5
+cpu_flags_arm_v6
+
+cpu_flags_arm_edsp
+cpu_flags_arm_thumb
+cpu_flags_arm_thumb2
+cpu_flags_arm_vfp
+cpu_flags_arm_vfp-d16
```
Comment (aballier) [inline profiles/arch/arm/armv6j/use.force]: Ditto, I don't think -d16 should be here

File: profiles/arch/arm/armv6j/use.force [hunk 11]
```diff
@@ -0,0 +1,10 @@
+# Force armv6 specific USE flags
+cpu_flags_arm_v4
+cpu_flags_arm_v5
+cpu_flags_arm_v6
+
+cpu_flags_arm_edsp
+cpu_flags_arm_thumb
+cpu_flags_arm_thumb2
+cpu_flags_arm_vfp
```
Comment (aballier) [inline profiles/arch/arm/armv6j/use.force]: https://www.arm.com/products/processors/technologies/vector-floating-point.php

VFPv2 is an optional extension to the ARM instruction set in the ARMv5TE, ARMv5TEJ and ARMv6 architectures.

Being optional it shouldnt be forced

File: profiles/arch/arm/armv6j/use.force [hunk 12]
```diff
@@ -0,0 +1,10 @@
+# Force armv6 specific USE flags
+cpu_flags_arm_v4
+cpu_flags_arm_v5
+cpu_flags_arm_v6
+
+cpu_flags_arm_edsp
+cpu_flags_arm_thumb
+cpu_flags_arm_thumb2
```
Comment (aballier) [inline profiles/arch/arm/armv6j/use.force]: armv6j doesn't seem to mandate thumb nor thumb2 (that'd be armv6t something)

File: profiles/arch/arm/armv7a/use.force [hunk 13]
```diff
@@ -2,3 +2,14 @@
...
+cpu_flags_arm_v4
+cpu_flags_arm_v5
+cpu_flags_arm_v6
+cpu_flags_arm_v7
+
+cpu_flags_arm_edsp
+cpu_flags_arm_thumb
+cpu_flags_arm_thumb2
+cpu_flags_arm_vfp
+cpu_flags_arm_vfpv3
```
Comment (aballier) [inline profiles/arch/arm/armv7a/use.force]: VFPv3 is an optional extension to the ARM, Thumb® and ThumbEE instruction sets in the ARMv7-A and ARMv7-R profiles.

File: profiles/arch/arm/armv5te/use.mask (hunk 9 above)
Comment (mgorny) [inline profiles/arch/arm/armv5te/use.mask]: I assume vfp means whatever the kernel means by vfp ;-). If you could figure that out, I'd appreciate.

File: profiles/arch/arm/armv6j/use.force (hunk 12 above)
Comment (mgorny) [inline profiles/arch/arm/armv6j/use.force]: Thumb seems to be mandatory since armv6, https://en.wikipedia.org/wiki/List_of_ARM_microarchitectures -- note that it's present even on plain ARMv6. About Thumb2, you're probably right.

File: profiles/desc/cpu_flags_arm.desc (hunk 8 above)
Comment (mgorny) [inline profiles/desc/cpu_flags_arm.desc]: I've used separate flags since that's how the kernel exposes it.

```
#define HWCAP_VFPv3	(1 << 13)
#define HWCAP_VFPv3D16	(1 << 14)	/* also set for VFPv4-D16 */
...
#define HWCAP_VFPv4	(1 << 16)
...
#define HWCAP_VFPD32	(1 << 19)	/* set if VFP has 32 regs (not 16) */
```

Comment (aballier) [inline profiles/desc/cpu_flags_arm.desc]: Well, how kernel and arm exposes this is one thing, what's most convenient for us is another one. They probably have reasons to expose it that way, but in that case I don't get them.

The way I understand it, the following required use would always hold:

vfpv3? ( ^^ ( vfp-d16 vfp-d32 ) )
vfp-d16? ( vfpv3 )
vfp-d32? ( vfpv3 )


So, in order to avoid having to write this required use line in every ebuild, to me it makes more sense to have:

vfpv3 - vfpv3 with at least 16 double registers
vfpv3-d32 - vfpv3 with at least 32 double registers

This should make all the cases expressible but avoids the need for required use.

File: profiles/arch/arm/armv5te/use.mask (hunk 9 above)
Comment (aballier) [inline profiles/arch/arm/armv5te/use.mask]: see above

File: profiles/desc/cpu_flags_arm.desc (hunk 8 above)
Comment (mgorny) [inline profiles/desc/cpu_flags_arm.desc]: Maybe just vfp-d32? Since it applies to v4 as well…

File: profiles/desc/cpu_flags_arm.desc (hunk 7 above)
Comment (mgorny) [inline profiles/desc/cpu_flags_arm.desc]: I haven't been able to figure it out from the kernel but chromeos assumes v2 here.

Comment (aballier) [inline profiles/desc/cpu_flags_arm.desc]: Since v1 is deprecated and apparently not publicly documented, I think it's safe to assume v2 too

The description should mention that I think

File: profiles/desc/cpu_flags_arm.desc (hunk 8 above)
Comment (aballier) [inline profiles/desc/cpu_flags_arm.desc]: Ah true, then vfp-d32 be it

**Comment by mgorny** [comment]
Ok, pushed an update to cpuid2cpuflags and the PR:
* `vfp-d16` has been removed,
* descriptions for `vfp*` have been improved,
* descriptions for `vN` flags have been improved to indicate that it's about instructions added in ARMvN, i.e. it should be on on newer versions.

**Comment by ZeroChaos-** [comment]
I thought that things like thumb and thumb2 were not to be enabled by default because there are trade-offs, but I'll yield to whatever @steev says.

**Comment by mgorny** [comment]
> I thought that things like thumb and thumb2 were not to be enabled by default because there are trade-offs, but I'll yield to whatever @steev says.

Well, the thumb* flags originate from the earlier armv5te & armv6t2 flags which were enabled this way. I'd say if upstream uses the flag (i.e. provides the option to use particular newer features), then there's a reason for it ;-).

**Comment by threeway** [comment]
In my opinion, thumb, thumb2, and neon should never be forced on.

Just because upstream uses a flag doesn't mean it's a good idea to turn it on.  For example, if your kernel isn't compiled in thumb2 mode, and you compile an external module in thumb2, you cannot load the module because they are incompatible.  There is a reason it's just a flag as an option, and having worked with a few upstreams, I can almost guarantee the only thing that was tested was that it compiled, not that it ran.

I also think we should specify vfpv3-d16/vfpv3-d32 or vfpv3-d16/vfpv3 - personally I'd prefer d16/d32 as it's more clear, but that's a personal preference and I won't lose sleep at all over whatever is chosen.

**Comment by steev** [comment]
(commenting from the correct account now)

I don't think thumb, thumb2 or neon should be forced on.   Just because upstream offers it doesn't mean it works, is well tested, or, even compatible.

A module built out of kernel for thumb2 won't work with a kernel that's thumb2.

I won't lose any sleep over it, but I'd prefer if we did vfpv3-d16 and vfpv3-d32 instead of just vfpv3 or vfpv3-d32.

I'd like it if @ssvb chimed in, but haven't seen him around in a while.

**Comment by mgorny** [comment]
> A module built out of kernel for thumb2 won't work with a kernel that's thumb2.

Is there a way to 'detect' whether kernel supports thumb2 out of userspace?

> I won't lose any sleep over it, but I'd prefer if we did vfpv3-d16 and vfpv3-d32 instead of just vfpv3 or vfpv3-d32.

it's vfpv3 [+ vfpv4] + vfp-d32 since -d16/-d32 can apply to vfpv4 as well.

**Comment by mgorny** [comment]
Oh wait, you were talking about kernel modules. Well, I'll defer that to @aballier since it's ffmpeg what uses those flags. We originally had e.g. armv6t2 but we figured out it's better to have just base versions + separate flags for extra thingies.

**Comment by aballier** [comment]
For ffmpeg it refers to instructions like 'sbfx' or 'movt' that have been added with armv6t2, that as I understand are the thumb2 instructions but those are also available/used in arm mode. I see the 'thumb2' flag as enabling support for those instructions, independently of whether we're building in arm or thumb mode. If you have any idea how to make this more clear, please share!
